File: Python/MCP3428.py
'''**************************************************************************/
Controleverything.com- MCP3428

https://www.controleverything.com/content/Analog-Digital-Converters?sku=MCP3428_I2CADC

Analog-to-Digital Converter with I2C Interface and On-Board Reference using Raspberry Pi

Firmware v1.0 - Python
Author - Amanpal Singh

4-Channel ADC with 16-Bit Resolution
4 Differential Analog Inputs
Programmable x1-x8 Gain Amplifier
Up to 8 Devices per I²C Port
Up to 3.4MHz Communication Speed
0x68 Start Address

Harware Version - Rev A.
Platform - Raspberry Pi

**************************************************************************'''

MCP3428_DEFAULT_ADDRESS	0x68

# MCP3422, MCP3423, MCP3424, MCP3427 & MCP3428 addresses are controlled by address lines A0 and A1
# each address line can be low (GND), high (VCC) or floating (FLT)
MCP3428_A0GND_A1GND		0x68
MCP3428_A0GND_A1FLT		0x69
MCP3428_A0GND_A1VCC		0x6A
MCP3428_A0FLT_A1GND		0x6B
MCP3428_A0VCC_A1GND		0x6C
MCP3428_A0VCC_A1FLT		0x6D
MCP3428_A0VCC_A1VCC		0x6E
MCP3428_A0FLT_A1VCC		0x6F

# Conversion mode definitions
MCP3428_MODE_ONESHOT	0x00
MCP3428_MODE_CONTINUOUS	0x10

# Channel definitions
# MCP3421 & MCP3425 have only the one channel and ignore this param
# MCP3422, MCP3423, MCP3426 & MCP3427 have two channels and treat 3 & 4 as repeats of 1 & 2 respectively
# MCP3424 & MCP3428 have all four channels
MCP3428_CHANNEL_1	0x00
MCP3428_CHANNEL_2	0x20
MCP3428_CHANNEL_3	0x40
MCP3428_CHANNEL_4	0x60
MCP3428_CHANNEL_MASK	0x60

# Sample size definitions - these also affect the sampling rate
# 12-bit has a max sample rate of 240sps
# 14-bit has a max sample rate of  60sps
# 16-bit has a max sample rate of  15sps
MCP3428_SIZE_12BIT	0x00
MCP3428_SIZE_14BIT	0x04
MCP3428_SIZE_16BIT	0x08
MCP3428_SIZE_18BIT	0x0C
MCP3428_SIZE_MASK	0x0C

# Programmable Gain definitions
MCP3428_GAIN_1X	0x00
MCP3428_GAIN_2X	0x01
MCP3428_GAIN_4X	0x02
MCP3428_GAIN_8X	0x03
MCP3428_GAIN_MASK 0x03

# /RDY bit definition
MCP3428_RDY	0x80

#IMporting the lib. and getting talk with I2C Bus.
import smbus
import time
bus = smbus.SMBus(1)


# Default :Channel 1,Sample Rate 240SPS(12- bit),Gain x1 Selected
def intialiseadc():
	bus.write_byte(0x68,0x80|0x10|0x00|0x00)

#Getting raw value from the device
def getadcreading() :
	data = bus.read_i2c_block_data(0x68,0x00,2)
	t = ((data[0] << 8) | data[1])
	print "Rawdata  : ",t
	if (t >= 32768):
		t = 65536 -t
	return t

#Conversion of the raw volatge accoring to the datasheet into volts
def getvoltage():
	
	v = (getadcreading() * 2.048)/2048	
	print "Voltage of the Source is : ", v,"volts"
		
	time.sleep(0.5)

#Calling the intialisation of the device
intialiseadc()

while True:
	
	#Getting the voltage of the source reading from the device
	getvoltage()

File: Java/MCP3428.java
/* Controleverything.com - MCP3428
 
 https://www.controleverything.com/content/Analog-Digital-Converters?sku=MCP3428_I2CADC
 
 Firmware v1.0 - java

 Author: Yadwinder Singh
 
 4-Channel ADC with 16-Bit Resolution
 4 Differential Analog Inputs
 Programmable x1-x8 Gain Amplifier
 Up to 8 Devices per I2C Port
 Up to 3.4MHz Communication Speed
 0x68 Start Address

 Hardware version - Rev A.
 
 Platform : Raspberry pi

 Project uses pi4j library. 
 Please follow a detailed tutorial to install pi4j here.
 
 http://pi4j.com/install.html
 
 Compile the java program with command ~\[Directory of code]\ pi4j Filename.java
 Run it with  ~\[Directory of code]\ pi4j Filename
 
 */


import java.nio.channels.WritableByteChannel;
import java.nio.ByteBuffer;

import com.pi4j.io.i2c.I2CBus;
import com.pi4j.io.i2c.I2CDevice;
import com.pi4j.io.i2c.I2CFactory;

import java.io.IOException;
import java.text.DecimalFormat;
import java.text.NumberFormat;
 

public class MCP3428
{
    public final static int MCP3428_ADDRESS = 0x68;
    
    public final static byte L = 0;
    public final static byte H = 1;
    public final static byte F = 2;
     
	
    public final static int CHANNEL_0 = 0;
    public final static int CHANNEL_1 = 1;
    public final static int CHANNEL_2 = 2;
    public final static int CHANNEL_3 = 3;
   
    public final static int GAIN_1 = 0;
    public final static int GAIN_2 = 1;
    public final static int GAIN_4 = 2;
    public final static int GAIN_8 = 3;

    double VRef = 2.048;  // Reference Voltage
     
    //communication register
    public final static int BIT_RDY = 7; //data ready
    public final static int BIT_C1 = 6; //channel select
    public final static int BIT_C0 = 5; //channel select
    public final static int BIT_OC = 4; //conversion mode (one shot/continuous)
    public final static int BIT_S1 = 3; //sample rate
    public final static int BIT_S0 = 2; //sample rate
    public final static int BIT_G1 = 1; //gain
    public final static int BIT_G0 = 0; //gain

    byte[] b = new byte[3];
    int status = 0;		
     
    private static boolean verbose = "true".equals(System.getProperty("mcp3428.verbose", "false"));
    
    private I2CBus bus;
    private I2CDevice mcp3428;
    
    public MCP3428()
    {
        this(MCP3428_ADDRESS);
    }
    
    public MCP3428(int address)
    {
        try
        {
            // Get i2c bus
            bus = I2CFactory.getInstance(I2CBus.BUS_1); // Depends onthe RasPI version
            if (verbose)
                System.out.println("Connected to bus. OK.");
            
            // Get device itself
            mcp3428 = bus.getDevice(address);
            if (verbose)
                System.out.println("Connected to device. OK.");
        }
        catch (IOException e)
        {
            System.err.println(e.getMessage());
        }
    }
    

   public void setConfig(int channel, int gain)             // Select Channel and Gain 
    {
        
        //configuration register, assuming 16 bit resolution
        byte reg =  (byte)((1 << BIT_RDY) | (channel << BIT_C0) | (1 << BIT_OC) | (1 << BIT_S1) | gain);
        
        try{
            
            mcp3428.write(reg);    // write configuration into config register.
        }
        catch (Exception ex)
        {
            System.err.println(ex.getMessage());
            ex.printStackTrace();
        }
        
        
    }
    public double readADC()
    
    {
        
        try{

		mcp3428.read(b,0,3);    // read byte data into byte buffer b
            	Thread.sleep(300);
            
        }
        catch (Exception ex)
        {
            System.err.println(ex.getMessage());
            ex.printStackTrace();
        }
        
        int h = b[0] & 0xff;
        int l = b[1] & 0xff;
        int r = b[2] & 0xff;
        
        long t = h << 8 |  l;
        
        
        if (t >= 32768)
            t = 65536l - t;
        

        System.out.println("ADC output :  " + t);  // print Raw ADC value
  
        double v = (double) t * VRef/32768.0;       // convert the ADC value voltage
        return v;
        
    }
    
    public static void main(String[] args)
    {
        
        MCP3428 sensor = new MCP3428();
	
        final NumberFormat NF = new DecimalFormat("##00.00");
        sensor.setConfig(CHANNEL_0,GAIN_1);     //selected Channel = 0 and Gain = 1x
        
	while(true)
        {
            
            System.out.println("OUTPUT FROM CHANNEL : " + sensor.readADC());
         
        }
        
        
    }
}

